[TypeChecker] Implement Double <-> CGFloat implicit conversion

xedin opened the pull request:
Implement an implicit conversion between `Double` and `CGFloat` types via
an implicit initializer call.

Resolves: rdar://problem/70592377

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

xedin committed:
[ConstraintSystem] Add a path element to represent an implicit conversion

xedin committed:
[AST] Add a helper to identify whether type is a CGFloat

xedin committed:
[ConstraintSystem] Add a new score kind to model implicit value conversions

xedin committed:
[ConstraintSystem] Add conversion restriction to cover implicit conversion from/to CGFloat

xedin committed:
[CSSimplify] Implement logic for an implicit conversion to/from CGFloat

To support implicit conversion to/from CGFloat type we need to:

- Reference an `init` member on the right-hand side type (either
  CGFloat if we are converting to it, or compatible type if from).

- Add a `ApplicableFunction` constraint to model an implicit
  call to previously referenced initializer with an argument
  of left-hand side type.

xedin committed:
[CSApply] Implement restriction-to-AST application to support implicit CGFloat conversion

Insert an implicit call to an initializer into the AST
to model to/from CGFloat conversion and pass converted
value as an argument (left-hand side of the conversion).

xedin committed:
[ConstraintSystem] Start using implicit conversion to/from CGFloat

Add logic to `matchTypes` to insert an implicit conversion to/from
CGFloat type when two nominal types are matched and one of them is
of `CGFloat` type.

xedin committed:
[CSStep] Make sure disjunction choices with value conversions are not considered "the best"

Augment `DisjunctionChoice::{shouldSkip, shouldStopAt}` to check
for presence of implicit value conversions, and if score indicates
that there is at least one of those was attempted - downgrade
importance of a last successful choice.

xedin committed:
[Diagnostics] NFC: Adjust a couple of tests which are now valid due to implicit CGFloat conversion

xedin committed:
[ConstraintSystem] Limit new implicit conversion to only Double <-> CGFloat

xedin committed:
[ConstraintLocator] Add `is` convenience method to simplify check for kind of implicit conversion

xedin committed:
[ConstraintSystem] Attempt Double <-> CGFloat implicit conversion only once and re-use results

xedin committed:
[ConstraintSystem] Simplify handling of implicit conversions in `getArgumentInfo`

xedin committed:
[TypeChecker] NFC: Add some more tests for Double <-> CGFloat implicit conversion

xedin commented:
@swift-ci please test source compatibility

xedin commented:
Debug build was a UPASS

xedin committed:
[ConstraintSystem] Favor overloads which match choice type exactly

xedin committed:
[TypeChecker] NFC: Add more situations to Double <-> CGFloat tests

benrimmington commented on include/swift/AST/Types.h:
```diff
@@ -790,7 +790,10 @@ class alignas(1 << TypeAlignInBits) TypeBase {
   /// Check if this is a nominal type defined at the top level of the Swift module
   bool isStdlibType();
-  
+
+  /// Check if this is a CGFloat type from `CoreGraphics` module.
+  bool isCGFloatType();
+
```
`CGFloat` also exists in open-source `Foundation`.

https://github.com/apple/swift-corelibs-foundation/blob/main/Sources/Foundation/CGFloat.swift

xedin replied:
Thanks! I'll need to check for both then and add some Linux tests.

xwu commented on test/Misc/misc_diagnostics.swift:
```diff
@@ -17,9 +17,10 @@ if let realRoomName = roomName as! NSString { // expected-warning{{forced cast f
 var pi = 3.14159265358979
 var d: CGFloat = 2.0
-var dpi:CGFloat = d*pi // expected-error@:21{{cannot convert value of type 'Double' to expected argument type 'CGFloat'}}
+var dpi:CGFloat = d*pi // Ok (implicit conversion Float -> CGFloat)
```
```suggestion
var dpi:CGFloat = d*pi // Ok (implicit conversion Double -> CGFloat)
```

xedin replied:
Thank you!

xwu commented:
This looks fantastic in terms of improving the usability of `CGFloat`.

To the extent that it touches upon the Swift language itself, and in particular type relationships and one of the original tentpole design decisions not to have implicit numeric conversions, I hope we'll have a chance to discuss and review this through the Swift Evolution process. If presented as a sort of built-in ABI-stable retroactive `typealias`-equivalent, it would be pretty much a no-brainer, but if presented as an "implicit conversion," what's implied (to some, at least) would be a much more radical departure from Swift's design.

xedin commented:
Thanks, @xwu! At this point I'm mostly trying to figure out whether it's feasible and what the performance impact is going to be. This would definitely be pitched and discussed on Swift Evolution if this experiment is a success.

xedin commented:
@swift-ci please build toolchain macOS

xedin committed:
[TypeChecker] NFC: Temporary mark test-case for rdar://46541800 as slow

xedin commented:
@swift-ci please build toolchain macOS

xedin committed:
[ConstraintSystem] Adjust impact of implicit Double <-> CGFloat conversions

- Prefer CGFloat -> Double over the other way around to avoid
  ambiguities;

- Every new conversion impacts the score by factor of number of
  previously applied conversions to make it possible to select
  solutions that require the least such conversions.

- Prefer concrete overloads with Double <-> CGFloat conversion
  over generic ones.

xedin committed:
[AST] Teach `isCGFloatType` that `CGFloat` could also be found in `Foundation`

xedin commented:
@swift-ci please build toolchain

xedin commented:
@swift-ci please test source compatibility

xedin commented:
@swift-ci please test source compatibility

xedin committed:
[AST] Add a new `isDoubleType` accessor to `TypeBase`

xedin committed:
[ConstraintSystem] Record all Double <-> CGFloat conversions performed along a current solver path

This information makes it easier to compute a score for the next
implicit value conversion solver would have to perform.

xedin committed:
[CSBindings] Prefer `Double` over `CGFloat` bindings

Always prefer binding to `Double` over `CGFloat` if possible
since there is an implicit conversion between them and attempting
both could lead to ambiguities.

xedin commented:
@swift-ci please test source compatibility

xedin commented:
@swift-ci please test source compatibility

xedin committed:
[CSApply] Make sure to load value before attempting implicit conversion

xedin commented:
@swift-ci please test source compatibility

xedin commented:
@swift-ci please build toolchain

xedin committed:
[stdlib] Revert changes to the CoreGraphics/CGFloat

xedin committed:
[CSRanking] Add `ImplicitValueConversion` to `getScoreKindName`

xedin committed:
[ConstraintSystem] Don't attempt Double-to-CGFloat in coercions/casts

xedin committed:
[ConstraintSystem] Increase impact of implicit Double<->CGFloat associated with contextual type

Consider conversion with context to be worse by default
because it means that expression as a whole couldn't get
to the expected type.

This also helps to disambiguate situations with
multiple solutions where one of them has a contextual
mismatch that requires implicit conversion and
another produces correct contextual type but requires
implicit conversion inside to get there.

xedin commented:
@swift-ci please test source compatibility

xedin committed:
[CSRanking] NFC: Fix bad merge artifact

xedin committed:
[ConstraintSystem] Allow Double<->CGFloat conversions only if they are AST representable

Allow Double <-> CGFloat conversion if the associated locator is
simplifiable down to a AST node (e.g. expression) this limits
applications of such conversions to e.g. argument-to-parameter,
contextual conversions and forbids conversions in generic arguments,
function result, collection elements etc.

xedin committed:
[CSBindings] Adjust Double<->CGFloat handling since `Bindings` are now a `SetVector`

xedin committed:
[ConstraintSystem] Increase impact of implicit value-to-value conversion

This makes it possible to pick `CGFloat` function/operator overloads even
in the presence of literals, otherwise non-default literal score gets in
a way and solver ends up producing a lot more solutions with implicit
Double<->CGFloat conversion than it can disambiguate, so it's better to
just preserve the old behavior and pick `CGFloat` concrete overloads when
appropriate.

xedin committed:
[CSGen] Favor CGFloat operator overloads for floating-pointer literal arguments

If argument is a floating-point literal, with newly introduced implicit
Double<->CGFloat conversion, sometimes it's better to choose a concrete
function/operator overload on `CGFloat` even if it's not a default literal
type e.g. `let _: CGFloat = min(1.0, log(<CGFloat value>))` shouldn't form
solutions with `Double` arguments since it would result in multiple
implicit conversions, it's better to use a `CGFloat` type for the arguments.

xedin committed:
[ConstraintSystem] Allow attempting generic unary operators in presence of Double<->CGFloat conversion

Not all of the unary operators have `CGFloat` overloads,
so in order to preserve previous behavior (and overall
best solution) with implicit Double<->CGFloat conversion
we need to allow  attempting generic operators for such cases.

```swift
let _: CGFloat = -.pi / 2
```

`-` doesn't have `CGFloat` overload (which might be an oversight),
so in order to preserve type-checking behavior solver can't be
allowed to pick `-(Double) -> Double` based on overload of `/`,
the best possible solution would be with `/` as `(CGFloat, CGFloat) -> CGFloat`
and `-` on a `FloatingPoint` protocol.

xedin commented:
@swift-ci please test source compatibility

xedin committed:
[TypeChecker] NFC: Add more tests for implicit Double<->CGFloat conversion

xedin committed:
[TypeChecker] Add a flag to disable Double<->CGFloat implicit conversion

xedin committed:
[CSBindings] Adopt Double <-> CGFloat handling to `addPotentialBindings` API change

xedin commented:
@swift-ci please build toolchain

xedin commented:
@swift-ci please build toolchain macOS

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please build toolchain macOS

xedin commented:
@swift-ci please test Windows platform

xedin commented:
@swift-ci please test macOS platform

xedin commented:
@swift-ci please test macOS platform

amayers commented on include/swift/AST/Types.h:
```diff
@@ -793,7 +793,13 @@ class alignas(1 << TypeAlignInBits) TypeBase {
+  /// Check if this is a CGFloat type from `CoreGraphics` module.
+  bool isCGFloatType();
+
+  /// Check if thih sis a Double type from standard library.
```
```suggestion
  /// Check if this is a Double type from standard library.
```

xedin commented:
@swift-ci please test macOS platform

xedin commented:
@swift-ci please build toolchain macOS

xedin commented:
@swift-ci please smoke test Windows platform

xedin commented:
@swift-ci please smoke test Windows platform

xedin committed:
[AST] NFC: Fix a typo in `TypeBase::isDoubleType` comment

xedin committed:
[CSSimplify] Avoid using designated initializer for `ArgumentInfo` (currently unsupported on Windows)

xedin committed:
[CSApply] Use correct locator while converting expression to contextual type

This wasn't a problem before since locator wasn't really used by
`ExprRewritter:coerceToType` but with Double<->CGFloat conversion
it needs the locator to be anchored at a rewritten expression instead
of the original one to form a correct implicit initializer call.

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test macOS platform

xedin commented:
@swift-ci please test Windows platform

xedin committed:
[CSGen] Favor overload choices that support Double/CGFloat conversions (argument or result)

xedin committed:
[ConstraintSystem] Adjust ranking rules associated with Double/CGFloat conversions

Change the conversion rule to favor any number of widenings (CGFloat -> Double)
over even a single narrowing conversion and if there is no way to avoid narrowing
(due to contextual requirements) attempt it as late as possible (the deeper in
the AST that conversion is located the higher its score).

This is a generally better rule when it comes to rounding and information
loss that would result from narrowing conversions.

xedin commented:
@swift-ci please test source compatibility

xedin committed:
[CSGen] Tweak favoring hacks to avoid favoring overloads that require narrowing conversions (Double -> CGFloat)

xedin commented:
@swift-ci please test source compatibility

xedin commented:
@swift-ci please test source compatibility release

xedin committed:
[CSGen] Don't favor overload choices that could require an implicit conversion

Doing so resulted in performance impact due to the number of partial
solutions increase.

xedin committed:
[CSSolver] Shrinking should avoid solutions with implicit conversions

Just like generic overloads, `shrink` should always avoid any solutions
with implicit conversions.

Reducing disjunction domains becaused on solutions with implicit
conversions could have negative performance impact due to the
increase in total number of solutions that have to be examined
as a result.

xedin commented:
@swift-ci please test source compatibility

xwu commented on include/swift/AST/Types.h:
```diff
@@ -799,7 +799,13 @@ class alignas(1 << TypeAlignInBits) TypeBase {
   bool isStdlibType();
-  
+
+  /// Check if this is a CGFloat type from `CoreGraphics` module.
```
Can you update the comment to reflect that you're also checking for this type in swift-corelibs-foundation?

xedin replied:
Sure!

xedin replied:
✅

xwu commented on include/swift/Basic/LangOptions.h:
```diff
@@ -579,6 +579,9 @@ namespace swift {
     /// See \ref FrontendOptions.PrintFullConvention
     bool PrintFullConvention = false;
+
+    /// Disallow Double and CGFloat types to be used interchangeably.
```
```suggestion
    /// Disallow Double and CGFloat types from being used interchangeably.
```

xedin replied:
✅

xwu commented on include/swift/Sema/Constraint.h:
```diff
@@ -258,6 +258,12 @@ enum class ConversionRestrictionKind {
+  /// an implicit CGFloat initializer call.
+  DoubleToCGFloat,
+  /// Implicit conversion from a value of CGFloat type to a value of Double type
+  /// via an implicit call to an initializer passing CGFloat value.
```
Is this just "via an implicit Double initializer call" or is the wording difference significant for some other nuance?

xedin replied:
I just wanted to point up what the argument is going to be in this case as well, since there are a lot of double initializer overloads out there.

xedin replied:
✅

xwu commented on include/swift/Sema/ConstraintSystem.h:
```diff
@@ -788,6 +792,10 @@ enum ScoreKind {
   SK_ForceUnchecked,
+  /// An implicit conversion from a value of one type (lhs)
+  /// to another type (rhs) via implicit initialization of
+  /// `rhs` type which an argument of `lhs` value.
```
```suggestion
  /// `rhs` type with an argument of `lhs` value.
```

xedin replied:
✅

xwu commented on include/swift/Sema/ConstraintSystem.h:
```diff
@@ -1393,6 +1405,10 @@ enum class ConstraintSystemFlags {
   /// `__attribute__((ns_consumed))`.
   UseClangFunctionTypes = 0x80,
+
+  /// Disallow to use Double and CGFloat interchangeably through means of
```
```suggestion
  /// Disallow using Double and CGFloat interchangeably by means of
```

xedin replied:
✅

xwu commented on lib/AST/Type.cpp:
```diff
@@ -812,6 +812,28 @@ bool TypeBase::isStdlibType() {
+    return false;
+
+  auto *module = DC->getParentModule();
+  // on macOS `CGFloat` is part of a `CoreGraphics` module,
```
```suggestion
  // On macOS `CGFloat` is part of a `CoreGraphics` module,
```

xedin replied:
@xwu I'm just going to lift this comment to the declaration as well to make it clear where `CGFloat` could be.

xedin replied:
✅

xwu commented on lib/Sema/CSGen.cpp:
```diff
@@ -544,15 +560,23 @@ namespace {
+                       ->getWithoutParens()
+                       ->getWithoutSpecifierType();
+
+      // There is no CGFloat overloads on some of the unary operators, so
```
```suggestion
      // There are no CGFloat overloads on some of the unary operators, so
```

xedin replied:
✅

xwu commented on lib/Sema/CSGen.cpp:
```diff
@@ -544,15 +560,23 @@ namespace {
+
+      // There is no CGFloat overloads on some of the unary operators, so
+      // in order to preserve current behavior let's not favor overloads
+      // which would result in conversion from CGFloat to Double otherwise
```
```suggestion
      // which would result in conversion from CGFloat to Double; otherwise
```

xedin replied:
✅

xwu commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -5089,6 +5083,70 @@ ConstraintSystem::matchTypes(Type type1, Type type2, ConstraintKind kind,
+
+        auto isCoercionOrCast = [](ASTNode anchor,
+                                   ArrayRef<LocatorPathElt> path) {
+          // e.g. contextual conversion from coercion/cast
```
```suggestion
          // E.g. contextual conversion from coercion/cast
```

xwu commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -10487,6 +10548,63 @@ ConstraintSystem::simplifyRestrictedConstraintImpl(
+    auto *memberLoc = getConstraintLocator(
+        applicationLoc, ConstraintLocator::ConstructorMember);
+
+    // This conversion has been already attempted for and constructor
```
Sorry, I think a word is missing here...

xedin replied:
✅

xwu commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -10487,6 +10548,63 @@ ConstraintSystem::simplifyRestrictedConstraintImpl(
+    if (findSelectedOverloadFor(memberLoc))
+      return SolutionKind::Solved;
+
+    // Allocate a single argument info to conver all possible
```
```suggestion
    // Allocate a single argument info to cover all possible
```

xedin replied:
✅

xwu commented on lib/Sema/CSStep.cpp:
```diff
@@ -696,8 +696,18 @@ bool DisjunctionStep::shouldSkip(const DisjunctionChoice &choice) const {
+    // Not all of the unary operators have `CGFloat` overloads,
+    // so in order to preserve previous behavior (and overall
+    // best solution) with implicit Double<->CGFloat conversion
+    // we need to allow  attempting generic operators for such cases.
```
```suggestion
    // we need to allow attempting generic operators for such cases.
```

xedin replied:
✅

xwu reviewed (commented):
I'm excited about this! Just some nits...

xedin committed:
[TypeChecker/SE-0307] NFC: Fix typos and clarify a couple of comments

Thanks Xiaodi Wu!

xedin commented:
@swift-ci please test source compatibility

xedin commented:
@swift-ci please test

xedin commented:
Windows failure:

```
Failed Tests (1):
  Swift(windows-x86_64) :: Concurrency/Runtime/async_task_detached.swift
```

xedin commented:
@swift-ci please test source compatibility release

xedin commented:
Release suite failed with `05:38:56 Caused: hudson.remoting.ChannelClosedException: Channel "hudson.remoting.Channel@79e5e9aa:macOS-21": Remote call on macOS-21 failed. The channel is closing down or has closed down`. Since SE-0307 got accepted I'm going to land this PR.
